compiler/circle-mpqsolver/src/bisection/ErrorApproximator.cpp

```diff
@@ -0,0 +1,361 @@
+
+inline bool has_min_max(const CircleNode *node)
+{
+  return node->quantparam() && !node->quantparam()->min.empty() && !node->quantparam()->max.empty();
+}
+
+inline uint32_t cal_offset(const loco::TensorShape &dimension, uint32_t *indices)
+{
+  return indices[0] * dimension.dim(1).value() * dimension.dim(2).value() *
+           dimension.dim(3).value() +
+         indices[1] * dimension.dim(2).value() * dimension.dim(3).value() +
+         indices[2] * dimension.dim(3).value() + indices[3];
+}
+
+uint32_t get_channel_dim_index(const CircleNode *node)
+{
+  uint32_t index = 0;
+  auto opcode = node->opcode();
+  switch (opcode)
+  {
+    case CircleOpcode::CONV_2D:
+    case CircleOpcode::TRANSPOSE_CONV:
+    case CircleOpcode::FULLY_CONNECTED:
+      index = 0;
+      break;
+    case CircleOpcode::DEPTHWISE_CONV_2D:
+      index = 3;
+      break;
+    default:
+      throw std::runtime_error("Failed to find channel index in " + node->name());
+  }
+
+  return index;
+}
+
+bool set_weight_dim(const CircleNode *node, const CircleConst *weights,
+                    loco::TensorShape &dimension)
+{
+  auto opcode = node->opcode();
+  switch (opcode)
+  {
+    case CircleOpcode::CONV_2D:
+    case CircleOpcode::TRANSPOSE_CONV:
+    case CircleOpcode::DEPTHWISE_CONV_2D:
+      assert(node->rank() == 4);
+      dimension.rank(node->rank());
+      dimension.dim(0).set(weights->dim(0).value());
+      dimension.dim(1).set(weights->dim(1).value());
+      dimension.dim(2).set(weights->dim(2).value());
+      dimension.dim(3).set(weights->dim(3).value());
+      break;
+    case CircleOpcode::FULLY_CONNECTED:
+      assert(node->rank() == 2);
+      dimension.rank(4);
+      dimension.dim(0).set(weights->dim(0).value());
+      dimension.dim(1).set(1); // Set FC layer like CONV
+      dimension.dim(2).set(1);
+      dimension.dim(3).set(weights->dim(1).value());
+      break;
+    default:
+      return false;
+  }
+
+  return true;
+}
+
+loco::Node *get_weight(const CircleNode *node)
+{
+  loco::Node *weight = nullptr;
+  auto opcode = node->opcode();
+  switch (opcode)
+  {
+    case CircleOpcode::CONV_2D:
+    {
+      auto conv = loco::must_cast<const CircleConv2D *>(node);
+      weight = conv->filter();
+    }
+    break;
+    case CircleOpcode::DEPTHWISE_CONV_2D:
+    {
+      auto dconv = loco::must_cast<const CircleDepthwiseConv2D *>(node);
+      weight = dconv->filter();
+    }
+    break;
+    case CircleOpcode::TRANSPOSE_CONV:
+    {
+      auto tconv = loco::must_cast<const CircleTransposeConv *>(node);
+      weight = tconv->filter();
+    }
+    break;
+    case CircleOpcode::FULLY_CONNECTED:
+    {
+      auto fc = loco::must_cast<const CircleFullyConnected *>(node);
+      weight = fc->weights();
+    }
+    break;
+    default:
+      break;
+  }
+
+  return weight;
+}
+
+void iterate_per_channel(const CircleNode *node, IterFunc func)
+{
+  CircleConst *weight = loco::must_cast<CircleConst *>(get_weight(node));
```

Comment:
Some models have non-const weights. Please throw an exception if weights is non-const.

Reply:
ok